handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -239,34 +239,36 @@ protected void deallocate() {
 
         Lock readerLock = context.ctxLock.readLock();
         readerLock.lock();
-        try {
-            ssl = SSL.newSSL(context.ctx, !context.isClient());
-        } finally {
-            readerLock.unlock();
-        }
-        try {
-            networkBIO = SSL.bioNewByteBuffer(ssl, context.getBioNonApplicationBufferSize());
+        synchronized (this) {
```

Comment:
Couldn't we have just merged the try blocks and used the reader lock? One thing that I don't understand is that you are adding sync inside the constructor, since the object cannot be accessed before the constructor has completed how can two threads race? On mobile so it's possible I've missed something..